zappa/core.py

```diff
@@ -511,20 +511,34 @@ def get_current_venv():
         """
         if "VIRTUAL_ENV" in os.environ:
             venv = os.environ["VIRTUAL_ENV"]
-        elif os.path.exists(".python-version"):  # pragma: no cover
-            try:
-                subprocess.check_output(["pyenv", "help"], stderr=subprocess.STDOUT)
-            except OSError:
-                print("This directory seems to have pyenv's local venv, " "but pyenv executable was not found.")
-            with open(".python-version", "r") as f:
-                # minor fix in how .python-version is read
-                # Related: https://github.com/Miserlou/Zappa/issues/921
-                env_name = f.readline().strip()
-            bin_path = subprocess.check_output(["pyenv", "which", "python"]).decode("utf-8")
-            venv = bin_path[: bin_path.rfind(env_name)] + env_name
-        else:  # pragma: no cover
-            return None
-        return venv
+            return venv
+
+        # pyenv available check
+        try: # progma: no cover
+            subprocess.check_output(["pyenv", "help"], stderr=subprocess.STDOUT)
+            pyenv_available = True
+        except OSError:
+            pyenv_available = False
+
+        if pyenv_available: # progma: no cover
+            # minor fix find pyenv version-file (.python-version)
+            # Related: https://github.com/zappa/Zappa/issues/1132
+            version_file_path = subprocess.check_output(
```

Comment:
`pyenv version-file` seems to return parent directory `.python-version` files if one doesn't exist in the current directory.

The current implementation appears to only look in the executed directory (?).
The `.python-version` file of the parent may not be the correct version.

Do we need to use the `.python-version` file?  Since we have the executable path, would something like this be better?

(could be cleaner, but just as an example)
```
...
env_name = subprocess.check_output([bin_path, "--version"]).strip().split(" ")[-1]
```

env_name appears to be the python version number, if that's the case maybe we should change the variable name to something like, `python_version`?

Reply:
Or even just parse the python version from the bin_path?

Reply:
The existing logic detects venv through pyenv only when a `.python-version` file exists in the current directory.

However, since pyenv can be used even if the `.python-version` file does not exist, the location of the two conditions has been changed. In other words, the `.python-version` directory can be used even if it does not exist in the current path.

In some projects, I've seen multiple repos use a single `.python-version` in the parent directory. However, it will still behave the same if the file is located in the current path as in the previous version.

Reply:
> Or even just parse the python version from the bin_path?
> 
> Is there a case where `pyenv which python` would give you the wrong version?
The existing pyenv detection logic is dependent on the `.python-version` file, so I implemented it with care only in this part. I couldn't find any problems with `pyenv which python`. However, it would be cool if the structure was changed as follows so that additional functions could be added later.


```python

        if pyenv_available:  # progma: no cover
            # minor fix find pyenv version-file (.python-version)
            # Related: https://github.com/zappa/Zappa/issues/1132
            version_file_path = subprocess.check_output(["pyenv", "version-file"]).decode("utf-8").strip()

            if ".python-version" in version_file_path:
                with open(version_file_path, "r") as f:
                    # minor fix in how .python-version is read
                    # Related: https://github.com/Miserlou/Zappa/issues/921
                    env_name = f.readline().strip()
                bin_path = subprocess.check_output(["pyenv", "which", "python"]).decode("utf-8")
                venv = bin_path[: bin_path.rfind(env_name)] + env_name
                return venv
                
            else: # bin_path
            # DO SOMETHING <<

        return None

```

Reply:
> env_name appears to be the python version number, if that's the case maybe we should change the variable name to something like, python_version?

In the .python-version file, the name of venv is often used in addition to the simple python version number. So, while I was thinking here too, I chose `env_name`, but if you want, I will change it to `python_version`.

Reply:
Yes, I understand that the existing code uses the `.python-version` file, but reviewing this pr and the pre-existing code, it looks like retrieving the version from the `.python-version` file is unnecessary.

It is only used as a _search string_ for correcting the bin_path:

```venv = bin_path[: bin_path.rfind(env_name)] + env_name```

In this case, it's isn't necessary to use the `.python-version`, and we can use the path discovered by `pyenv which python`, using the right most match.

Could we use something like this to bypass the need to use the `.python-version` file?

```
import re

python_version_match_regex = re.compile("(3\.\d+\.\d+)")

p = "/home/3.8.1/sample/.pyenv/versions/3.8.5/bin/python"
matches = [m for m in python_version_match_regex.finditer(p)]  # get all matches
match_count = len(matches)
last_match = matches[-1]
_, last_index = last_match.span()  # get position of last match
venv = p[:last_index]
```

Reply:
If I understood correctly, I have one concern.

This is because, in the `.python-version` file, various virtual environment names specified by the user can be specified in addition to the simple python version. For example, the example I am using is as follows:

```
> pwd
/Users/imiml/Documents/GitHub/imfast

> pyenv version-file
/Users/imiml/Documents/GitHub/imfast/.python-version

> cat /Users/imiml/Documents/GitHub/imfast/.python-version
imfast

> pyenv which python
/Users/imiml/.pyenv/versions/imfast/bin/python

```

For this reason, I think that finding it by pattern matching consisting of simple numbers is risky.

As an alternative, it may be possible if it matches and removes like `/bin/python`. However, there may be other exceptions, so this is also a concern.

Reply:
Interesting, I wasn't aware of that.
Yes, we might be able to remove just the last `/bin/python`.  I'm not sure was well if that's stable.

Reply:
From the pyenv documentation:

> Each Python version is installed into its own directory under $(pyenv root)/versions

We could get the value after `versions` and the system separator may be.

Reply:
this may work:
```
$ pyenv version-name
3.9.6
```

Reply:
something like this?
```
pyenv_root = subprocess.check_output(["pyenv", "root"]).decode("utf-8").strip()
pyenv_version = subprocess.check_output(["pyenv", "version-name"]).decode("utf-8").strip()
venv = str(Path(pyenv_root)) / pyenv_version
```

Reply:
wow, that's a good idea I will implement it soon.